File: python/signalhouse/domains/messages.py
"""Messages domain for the SignalHouse SDK."""

from __future__ import annotations

import json
from typing import Any, BinaryIO, TYPE_CHECKING
from urllib.parse import quote

if TYPE_CHECKING:
    from ..client import SignalHouseSDK


class Messages:
    """SMS/MMS messaging operations with multipart file upload support."""

    def __init__(self, sdk: SignalHouseSDK) -> None:
        self._sdk = sdk

    def get_messages(
        self,
        *,
        id: str | None = None,
        campaign_id: str | None = None,
        brand_id: str | None = None,
        subgroup_id: str | None = None,
        group_id: str | None = None,
        phone_number: str | None = None,
        status: str | None = None,
        direction: str | None = None,
        message_type: str | None = None,
        carrier: str | None = None,
        start_date: str | None = None,
        end_date: str | None = None,
        sort_field: str | None = None,
        sort_order: str | None = None,
        page: int | None = None,
        limit: int | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get a list of messages with optional filters and pagination.

        Args:
            id: Filter messages by their unique ID.
            campaign_id: Filter messages by the ID of the associated campaign.
            brand_id: Filter messages by the ID of the associated brand.
            subgroup_id: Filter messages by the ID of the associated subgroup.
            group_id: Filter messages by the ID of the associated group.
            status: Filter by status (ENQUEUED, DEQUEUED, SENT, FAILED, DELIVERED).
            direction: Filter by direction (INBOUND, OUTBOUND).
            message_type: Filter by type (SMS, MMS).
            carrier: Filter by carrier used for sending.
            start_date: Filter by start date.
            end_date: Filter by end date.
            sort_field: The field to sort by.
            sort_order: The sort order (asc, desc).
            page: The page number for pagination.
            limit: The number of messages per page.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.
        """
        query_string = self._sdk._get_query_string({
            "id": id,
            "campaignId": campaign_id,
            "brandId": brand_id,
            "subgroupId": subgroup_id,
            "groupId": group_id,
            "phoneNumber": phone_number,
            "status": status,
            "direction": direction,
            "messageType": message_type,
            "carrier": carrier,
            "startDate": start_date,
            "endDate": end_date,
            "sortField": sort_field,
            "sortOrder": sort_order,
            "page": page,
            "limit": limit,
        })
        return self._sdk._request(
            f"/message{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def get_analytics(
        self,
        *,
        group_id: str | None = None,
        subgroup_id: str | None = None,
        brand_id: str | None = None,
        campaign_id: str | None = None,
        phone_number: str | None = None,
        carrier: str | None = None,
        start_date: str | None = None,
        end_date: str | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get aggregated analytics for messages with optional filters.

        Args:
            group_id: Filter analytics by group ID.
            subgroup_id: Filter analytics by subgroup ID.
            brand_id: Filter analytics by brand ID.
            campaign_id: Filter analytics by campaign ID.
            phone_number: Filter analytics by phone number.
            carrier: Filter analytics by carrier.
            start_date: Filter analytics by start date.
            end_date: Filter analytics by end date.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.
        """
        query_string = self._sdk._get_query_string({
            "groupId": group_id,
            "subgroupId": subgroup_id,
            "brandId": brand_id,
            "campaignId": campaign_id,
            "phoneNumber": phone_number,
            "carrier": carrier,
            "startDate": start_date,
            "endDate": end_date,
        })
        return self._sdk._request(
            f"/message/analytics{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def get_analytics_detail(
        self,
        *,
        group_id: str | None = None,
        subgroup_id: str | None = None,
        brand_id: str | None = None,
        campaign_id: str | None = None,
        phone_number: str | None = None,
        carrier: str | None = None,
        start_date: str | None = None,
        end_date: str | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get detailed analytics snapshot records for charting and aggregation.

        Args:
            group_id: Filter analytics by group ID.
            subgroup_id: Filter analytics by subgroup ID.
            brand_id: Filter analytics by brand ID.
            campaign_id: Filter analytics by campaign ID.
            phone_number: Filter analytics by phone number.
            carrier: Filter analytics by carrier.
            start_date: Filter analytics by start date.
            end_date: Filter analytics by end date.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict with array of analytics snapshot records.
        """
        query_string = self._sdk._get_query_string({
            "groupId": group_id,
            "subgroupId": subgroup_id,
            "brandId": brand_id,
            "campaignId": campaign_id,
            "phoneNumber": phone_number,
            "carrier": carrier,
            "startDate": start_date,
            "endDate": end_date,
        })
        return self._sdk._request(
            f"/message/analytics/detail{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def get_dnc_analytics(
        self,
        *,
        group_id: str | None = None,
        subgroup_id: str | None = None,
        brand_id: str | None = None,
        campaign_id: str | None = None,
        phone_number: str | None = None,
        carrier: str | None = None,
        start_date: str | None = None,
        end_date: str | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get aggregated DNC (Do Not Contact) opt-out analytics with optional filters.

        Args:
            group_id: Filter by group ID.
            subgroup_id: Filter by subgroup ID.
            brand_id: Filter by brand ID.
            campaign_id: Filter by campaign ID.
            phone_number: Filter by phone number.
            carrier: Filter by carrier.
            start_date: Filter by start date.
            end_date: Filter by end date.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict with totals, byDate, byPhoneNumber, byCarrier.
        """
        query_string = self._sdk._get_query_string({
            "groupId": group_id,
            "subgroupId": subgroup_id,
            "brandId": brand_id,
            "campaignId": campaign_id,
            "phoneNumber": phone_number,
            "carrier": carrier,
            "startDate": start_date,
            "endDate": end_date,
        })
        return self._sdk._request(
            f"/message/dnc/analytics{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def get_dnc_records(
        self,
        *,
        group_id: str | None = None,
        subgroup_id: str | None = None,
        brand_id: str | None = None,
        campaign_id: str | None = None,
        phone_number: str | None = None,
        carrier: str | None = None,
        start_date: str | None = None,
        end_date: str | None = None,
        page: int | None = None,
        limit: int | None = None,
        sort_field: str | None = None,
        sort_order: str | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get paginated Do Not Call records with optional filters.

        Args:
            group_id: Filter by group ID.
            subgroup_id: Filter by subgroup ID.
            brand_id: Filter by brand ID.
            campaign_id: Filter by campaign ID.
            phone_number: Filter by phone number.
            carrier: Filter by carrier.
            start_date: Filter by start date.
            end_date: Filter by end date.
            page: Page number for pagination.
            limit: Number of records per page.
            sort_field: Field to sort by.
            sort_order: Sort direction (asc, desc).
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict with paginated DNC records.
        """
        query_string = self._sdk._get_query_string({
            "groupId": group_id,
            "subgroupId": subgroup_id,
            "brandId": brand_id,
            "campaignId": campaign_id,
            "phoneNumber": phone_number,
            "carrier": carrier,
            "startDate": start_date,
            "endDate": end_date,
            "page": page,
            "limit": limit,
            "sortField": sort_field,
            "sortOrder": sort_order,
        })
        return self._sdk._request(
            f"/message/dnc{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def send_sms(
        self,
        sender_phone_number: str,
        recipient_phone_numbers: str | list[str],
        message_body: str,
        *,
        status_callback_url: str | None = None,
        enable_shortlink: bool = False,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Send an SMS message to one or more recipient phone numbers.

        Args:
            sender_phone_number: The phone number to send the message from.
            recipient_phone_numbers: The phone number(s) to send the message to.
            message_body: The body of the SMS message.
            status_callback_url: The URL to receive status callbacks.
            enable_shortlink: Whether to enable shortlink in the message.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If required parameters are missing.
        """
        self._sdk._require({
            "senderPhoneNumber": sender_phone_number,
            "recipientPhoneNumbers": recipient_phone_numbers,
            "messageBody": message_body,
        })
        body: dict[str, Any] = {
            "senderPhoneNumber": sender_phone_number,
            "recipientPhoneNumber": recipient_phone_numbers,
            "messageBody": message_body,
            "enableShortlink": enable_shortlink,
        }
        if status_callback_url is not None:
            body["statusCallbackUrl"] = status_callback_url
        return self._sdk._request(
            "/message/sms",
            method="POST",
            body=body,
            token=token,
            headers=headers,
        )

    def send_mms(
        self,
        sender_phone_number: str,
        recipient_phone_numbers: str | list[str],
        message_body: str,
        *,
        media_urls: list[str] | None = None,
        status_callback_url: str | None = None,
        enable_shortlink: bool = False,
        enable_compression: bool = True,
        images: list[BinaryIO | tuple] | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Send an MMS message to one or more recipients, with optional media attachments.

        Args:
            sender_phone_number: The phone number to send the message from.
            recipient_phone_numbers: The phone number(s) to send the message to.
            message_body: The body of the MMS message.
            media_urls: The URLs of the media attachments.
            status_callback_url: The URL to receive status callbacks.
            enable_shortlink: Whether to enable shortlink in the message.
            enable_compression: Whether to enable image compression for attachments.
            images: Image files to attach to the MMS message. Each can be a file-like object
                    or a tuple of (filename, file_object, content_type).
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If required parameters are missing.
        """
        self._sdk._require({
            "senderPhoneNumber": sender_phone_number,
            "recipientPhoneNumbers": recipient_phone_numbers,
            "messageBody": message_body,
        })

        form_data: dict[str, Any] = {
            "senderPhoneNumber": sender_phone_number,
            "messageBody": message_body,
            "enableShortlink": str(enable_shortlink).lower(),
            "enableCompression": str(enable_compression).lower(),
        }

        if isinstance(recipient_phone_numbers, list):
            form_data["recipientPhoneNumber"] = json.dumps(recipient_phone_numbers)
        else:
            form_data["recipientPhoneNumber"] = recipient_phone_numbers

        if media_urls:
            form_data["mediaUrls"] = json.dumps(media_urls)

        if status_callback_url:
            form_data["statusCallbackUrl"] = status_callback_url

        files_list: list[tuple[str, Any]] = []
        if images:
            for image in images:
                if isinstance(image, tuple):
                    files_list.append(("images", image))
                else:
                    files_list.append(("images", image))

        return self._sdk._multipart_request(
            "/message/mms",
            method="POST",
            form_data=form_data,
            files=files_list if files_list else None,
            token=token,
            headers=headers,
        )

    def send_group_message(
        self,
        sender_phone_number: str,
        recipient_phone_numbers: str | list[str],
        message_body: str,
        *,
        media_urls: list[str] | None = None,
        status_callback_url: str | None = None,
        enable_shortlink: bool = False,
        enable_compression: bool = True,
        images: list[BinaryIO | tuple] | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Send a group MMS message to one or more recipients, with optional media attachments.

        Args:
            sender_phone_number: The phone number to send the message from.
            recipient_phone_numbers: The phone number(s) to send the message to.
            message_body: The body of the MMS message.
            media_urls: The URLs of the media attachments.
            status_callback_url: The URL to receive status callbacks.
            enable_shortlink: Whether to enable shortlink in the message.
            enable_compression: Whether to enable image compression for attachments.
            images: Image files to attach to the MMS message. Each can be a file-like object
                    or a tuple of (filename, file_object, content_type).
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If required parameters are missing.
        """
        self._sdk._require({
            "senderPhoneNumber": sender_phone_number,
            "recipientPhoneNumbers": recipient_phone_numbers,
            "messageBody": message_body,
        })

        form_data: dict[str, Any] = {
            "senderPhoneNumber": sender_phone_number,
            "messageBody": message_body,
            "enableShortlink": str(enable_shortlink).lower(),
            "enableCompression": str(enable_compression).lower(),
        }

        if isinstance(recipient_phone_numbers, list):
            form_data["recipientPhoneNumber"] = json.dumps(recipient_phone_numbers)
        else:
            form_data["recipientPhoneNumber"] = recipient_phone_numbers

        if media_urls:
            form_data["mediaUrls"] = json.dumps(media_urls)

        if status_callback_url:
            form_data["statusCallbackUrl"] = status_callback_url

        files_list: list[tuple[str, Any]] = []
        if images:
            for image in images:
                if isinstance(image, tuple):
                    files_list.append(("images", image))
                else:
                    files_list.append(("images", image))

        return self._sdk._multipart_request(
            "/message/groupMessage",
            method="POST",
            form_data=form_data,
            files=files_list if files_list else None,
            token=token,
            headers=headers,
        )

File: php/src/Domains/Messages.php
<?php

namespace SignalHouse\SDK\Domains;

use SignalHouse\SDK\HttpClient;

class Messages
{
    private HttpClient $client;
    private HttpClient $multipartClient;
    private bool $enableAdmin;

    public function __construct(HttpClient $client, HttpClient $multipartClient, bool $enableAdmin)
    {
        $this->client = $client;
        $this->multipartClient = $multipartClient;
        $this->enableAdmin = $enableAdmin;
    }

    /**
     * Get a list of messages with optional filters and pagination
     *
     * @param array $params Filter parameters (id, campaignId, brandId, subgroupId, groupId, status, direction, messageType, carrier, startDate, endDate, sortField, sortOrder, page, limit)
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function getMessages(array $params = [], array $options = []): array
    {
        $queryString = $this->client->getQueryString($params);
        return $this->client->request("/message{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Get aggregated analytics for messages with optional filters
     *
     * @param array $params Filter parameters (groupId, subgroupId, brandId, campaignId, phoneNumber, carrier, startDate, endDate)
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function getAnalytics(array $params = [], array $options = []): array
    {
        $queryString = $this->client->getQueryString($params);
        return $this->client->request("/message/analytics{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Get detailed analytics snapshot records for charting and aggregation
     *
     * @param array $params Filter parameters (groupId, subgroupId, brandId, campaignId, phoneNumber, carrier, startDate, endDate)
     * @param array $options Additional request options
     * @return array The response from the server with an array of analytics snapshot records
     */
    public function getAnalyticsDetail(array $params = [], array $options = []): array
    {
        $queryString = $this->client->getQueryString($params);
        return $this->client->request("/message/analytics/detail{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Get aggregated DNC (Do Not Contact) opt-out analytics with optional filters
     *
     * @param array $params Filter parameters (groupId, subgroupId, brandId, campaignId, phoneNumber, carrier, startDate, endDate)
     * @param array $options Additional request options
     * @return array The response from the server with totals, byDate, byPhoneNumber, byCarrier
     */
    public function getDncAnalytics(array $params = [], array $options = []): array
    {
        $queryString = $this->client->getQueryString($params);
        return $this->client->request("/message/dnc/analytics{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Get paginated Do Not Call records with optional filters
     *
     * @param array $params Filter parameters (groupId, subgroupId, brandId, campaignId, phoneNumber, carrier, startDate, endDate, page, limit, sortField, sortOrder)
     * @param array $options Additional request options
     * @return array The response from the server with paginated DNC records
     */
    public function getDncRecords(array $params = [], array $options = []): array
    {
        $queryString = $this->client->getQueryString($params);
        return $this->client->request("/message/dnc{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Send an SMS message
     *
     * @param string $senderPhoneNumber The phone number to send from
     * @param string|array $recipientPhoneNumbers The recipient phone number(s)
     * @param string $messageBody The message body
     * @param string|null $statusCallbackUrl Optional callback URL for status updates
     * @param bool $enableShortlink Whether to enable shortlinks
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function sendSMS(
        string $senderPhoneNumber,
        string|array $recipientPhoneNumbers,
        string $messageBody,
        ?string $statusCallbackUrl = null,
        bool $enableShortlink = false,
        array $options = []
    ): array {
        $this->client->require([
            'senderPhoneNumber' => $senderPhoneNumber,
            'recipientPhoneNumbers' => $recipientPhoneNumbers,
            'messageBody' => $messageBody,
        ]);

        $body = [
            'senderPhoneNumber' => $senderPhoneNumber,
            'recipientPhoneNumber' => $recipientPhoneNumbers,
            'messageBody' => $messageBody,
            'enableShortlink' => $enableShortlink,
        ];

        if ($statusCallbackUrl !== null) {
            $body['statusCallbackUrl'] = $statusCallbackUrl;
        }

        return $this->client->request('/message/sms', array_merge([
            'method' => 'POST',
            'body' => $body,
        ], $options));
    }

    /**
     * Send an MMS message with optional media attachments
     *
     * @param string $senderPhoneNumber The phone number to send from
     * @param array $recipientPhoneNumbers The recipient phone numbers
     * @param string $messageBody The message body
     * @param array|null $mediaUrls Optional media URLs
     * @param string|null $statusCallbackUrl Optional callback URL
     * @param bool $enableShortlink Whether to enable shortlinks
     * @param bool $enableCompression Whether to enable image compression
     * @param array|null $images Optional image file paths
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function sendMMS(
        string $senderPhoneNumber,
        array $recipientPhoneNumbers,
        string $messageBody,
        ?array $mediaUrls = null,
        ?string $statusCallbackUrl = null,
        bool $enableShortlink = false,
        bool $enableCompression = true,
        ?array $images = null,
        array $options = []
    ): array {
        $this->client->require([
            'senderPhoneNumber' => $senderPhoneNumber,
            'recipientPhoneNumbers' => $recipientPhoneNumbers,
            'messageBody' => $messageBody,
        ]);

        $multipart = [];

        if ($images !== null) {
            foreach ($images as $image) {
                $multipart[] = [
                    'name' => 'images',
                    'contents' => is_string($image) ? fopen($image, 'r') : $image,
                ];
            }
        }

        $multipart[] = ['name' => 'senderPhoneNumber', 'contents' => $senderPhoneNumber];
        $multipart[] = ['name' => 'recipientPhoneNumber', 'contents' => json_encode($recipientPhoneNumbers)];
        $multipart[] = ['name' => 'messageBody', 'contents' => $messageBody];

        if ($mediaUrls !== null && count($mediaUrls) > 0) {
            $multipart[] = ['name' => 'mediaUrls', 'contents' => json_encode($mediaUrls)];
        }

        if ($statusCallbackUrl !== null) {
            $multipart[] = ['name' => 'statusCallbackUrl', 'contents' => $statusCallbackUrl];
        }

        $multipart[] = ['name' => 'enableShortlink', 'contents' => $enableShortlink ? 'true' : 'false'];
        $multipart[] = ['name' => 'enableCompression', 'contents' => $enableCompression ? 'true' : 'false'];

        return $this->multipartClient->request('/message/mms', array_merge([
            'method' => 'POST',
            'multipart' => $multipart,
        ], $options));
    }

    /**
     * Send a group MMS message
     *
     * @param string $senderPhoneNumber The phone number to send from
     * @param array $recipientPhoneNumbers The recipient phone numbers
     * @param string $messageBody The message body
     * @param array|null $mediaUrls Optional media URLs
     * @param string|null $statusCallbackUrl Optional callback URL
     * @param bool $enableShortlink Whether to enable shortlinks
     * @param bool $enableCompression Whether to enable image compression
     * @param array|null $images Optional image file paths
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function sendGroupMessage(
        string $senderPhoneNumber,
        array $recipientPhoneNumbers,
        string $messageBody,
        ?array $mediaUrls = null,
        ?string $statusCallbackUrl = null,
        bool $enableShortlink = false,
        bool $enableCompression = true,
        ?array $images = null,
        array $options = []
    ): array {
        $this->client->require([
            'senderPhoneNumber' => $senderPhoneNumber,
            'recipientPhoneNumbers' => $recipientPhoneNumbers,
            'messageBody' => $messageBody,
        ]);

        $multipart = [];

        if ($images !== null) {
            foreach ($images as $image) {
                $multipart[] = [
                    'name' => 'images',
                    'contents' => is_string($image) ? fopen($image, 'r') : $image,
                ];
            }
        }

        $multipart[] = ['name' => 'senderPhoneNumber', 'contents' => $senderPhoneNumber];
        $multipart[] = ['name' => 'recipientPhoneNumber', 'contents' => json_encode($recipientPhoneNumbers)];
        $multipart[] = ['name' => 'messageBody', 'contents' => $messageBody];

        if ($mediaUrls !== null && count($mediaUrls) > 0) {
            $multipart[] = ['name' => 'mediaUrls', 'contents' => json_encode($mediaUrls)];
        }

        if ($statusCallbackUrl !== null) {
            $multipart[] = ['name' => 'statusCallbackUrl', 'contents' => $statusCallbackUrl];
        }

        $multipart[] = ['name' => 'enableShortlink', 'contents' => $enableShortlink ? 'true' : 'false'];
        $multipart[] = ['name' => 'enableCompression', 'contents' => $enableCompression ? 'true' : 'false'];

        return $this->multipartClient->request('/message/groupMessage', array_merge([
            'method' => 'POST',
            'multipart' => $multipart,
        ], $options));
    }
}
